### PR TITLE
Make mon_info_t consistent with mon_addr

### PR DESCRIPTION
**Pull request by jshen28**
Using the same logic for constructing MonMap::mon_info and mon_addr.

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

**Comment by jshen28** [comment]
jenkins test api

File: PendingReleaseNotes [hunk 1]
```diff
@@ -35,6 +35,13 @@
   users that modifying "max_mds" may not help with troubleshooting or recovery
   effort. Instead, it might further destabilize the cluster.
 
+* mgr/restful, mgr/zabbix: both modules, already deprecated since 2020, have been
+  finally removed. They have not been actively maintenance in the last years,
+  and started suffering from vulnerabilities in their dependency chain (e.g.:
+  CVE-2023-46136).  As alternatives, for the `restful` module, the `dashboard` module
+  provides a richer and better maintained RESTful API. Regarding the `zabbix` module,
+  there are alternative monitoring solutions, like `prometheus`, which is the most
```
Comment (anthonyeleven) [inline PendingReleaseNotes:43]: Prometheus

**Review by anthonyeleven — COMMENTED** [review]
Many unrelated files here.  Rebase gone wild?

**Comment by zdover23** [comment]
@jshen28, Separate out all docs changes as described in the "Note" section of this:  https://docs.ceph.com/en/reef/start/documenting-ceph/#select-a-branch

**Comment by jshen28** [comment]
yes, let me fix it...

File: PendingReleaseNotes [hunk 2]
```diff
@@ -35,6 +35,13 @@
   users that modifying "max_mds" may not help with troubleshooting or recovery
   effort. Instead, it might further destabilize the cluster.
 
+* mgr/restful, mgr/zabbix: both modules, already deprecated since 2020, have been
+  finally removed. They have not been actively maintenance in the last years,
```
[the remark this answers is not in the document]
Comment (jshen28) [reply PendingReleaseNotes]: rebase fixed.. thank you

**Comment by jshen28** [comment]
thank you. I wrongly rebased from main branch and include many unrelated changes..

File: src/mon/MonMap.cc [hunk 3]
```diff
@@ -196,7 +196,11 @@ void MonMap::encode(ceph::buffer::list& blist, uint64_t con_features) const
   if (!HAVE_FEATURE(con_features, MONENC) ||
       !HAVE_FEATURE(con_features, SERVER_NAUTILUS)) {
     for (auto& [name, info] : mon_info) {
-      legacy_mon_addr[name] = info.public_addrs.legacy_addr();
+      auto addr = info.public_addrs.legacy_addr();
+      if ( addr == entity_addr_t()) {
```
Comment (idryomov) [inline src/mon/MonMap.cc:200]:
```suggestion
      if (addr == entity_addr_t()) {
```

File: src/mon/MonMap.cc [hunk 4]
```diff
@@ -196,7 +196,11 @@ void MonMap::encode(ceph::buffer::list& blist, uint64_t con_features) const
   if (!HAVE_FEATURE(con_features, MONENC) ||
       !HAVE_FEATURE(con_features, SERVER_NAUTILUS)) {
     for (auto& [name, info] : mon_info) {
-      legacy_mon_addr[name] = info.public_addrs.legacy_addr();
+      auto addr = info.public_addrs.legacy_addr();
```
Comment (idryomov) [inline src/mon/MonMap.cc:199]: Add a comment saying
```
// see note in mon_info_t::encode()
```
or something like that above this line.

**Review by idryomov — COMMENTED** [review]
Overall LGTM.  This was obviously missed in https://github.com/ceph/ceph/pull/31472.

**Comment by idryomov** [comment]
This fix came out of https://tracker.ceph.com/issues/68743.

File: src/mon/MonMap.cc (hunk 3 above)
Comment (jshen28) [reply src/mon/MonMap.cc:200]: thank you, fixed.

File: src/mon/MonMap.cc (hunk 4 above)
Comment (jshen28) [reply src/mon/MonMap.cc:199]: comment added.

File: src/mon/MonMap.cc (hunk 3 above)
Comment (idryomov) [reply src/mon/MonMap.cc:200]: I don't see it, have you pushed the fix?

Comment (jshen28) [reply src/mon/MonMap.cc]: Ah sorry, forget to push the updated patch...

**Review by idryomov — APPROVED** [review]

**Comment by jshen28** [comment]
gentle ping

**Comment by idryomov** [comment]
I built a pre-nautilus client (Ceph 12.2.13) and tested this manually against a cluster built on main where:

- monmap looked like this:
```
0: [v2:192.168.0.1:41100/0,v1:192.168.0.1:41101/0] mon.a
1: v2:192.168.0.1:41103/0 mon.b
2: v2:192.168.0.1:41105/0 mon.c
```
- ceph.conf file for the client had this:
```
mon host = 192.168.0.1:41101 192.168.0.1:41103 192.168.0.1:41105
```
- `auth_allow_insecure_global_id_reclaim` was set to true

In this setup the encoded monmap would misrepresent/lie about mon.b and mon.c addresses.

Without this PR, `ceph status` (or any other) command crashed as in https://tracker.ceph.com/issues/68743.  With this PR, it worked (sometimes expectedly taking a bit of time to establish the monitor session -- when mon.b and mon.c addresses got randomly chosen as the initial pair of monitors to connect to).  An older kernel client that doesn't advertise support for `MONENC` and `SERVER_NAUTILUS` (kernel 5.9) worked too.

**Pull request merged**